src/Flysystem/FlysystemAssetStore.php

```diff
@@ -913,24 +954,94 @@ public static function flush()
 
     public function getResponseFor($asset)
     {
-        // Check if file exists
-        $filesystem = $this->getFilesystemFor($asset);
-        if (!$filesystem) {
-            return $this->createMissingResponse();
+
+        $public = $this->getPublicFilesystem();
+        $protected = $this->getProtectedFilesystem();
+
+        // If the file exists on the public store, we just straight return it.
+        if ($public->has($asset)) {
+            return $this->createResponseFor($public, $asset);
         }
 
-        // Block directory access
-        if ($filesystem->get($asset) instanceof Directory) {
-            return $this->createDeniedResponse();
+        // If the file exists in the protected store and the user has been explicitely granted access to it
+        if ($protected->has($asset) && $this->isGranted($asset)) {
+            return $this->createResponseFor($protected, $asset);
+            // Let's not deny if the file is in the protected store, but is not granted.
+            // We might be able to redirect to a live version.
+        }
+
+        // If we found a URL to redirect to
+        if ($redirectUrl = $this->searchForEquivalentFileID($asset)) {
+            if ($redirectUrl != $asset && $public->has($redirectUrl)) {
+                return $this->createRedirectResponse($redirectUrl);
+            } else {
+                // Something weird is going on e.g. a publish file without a physical file
+                return $this->createMissingResponse();
+            }
         }
 
         // Deny if file is protected and denied
-        if ($filesystem === $this->getProtectedFilesystem() && !$this->isGranted($asset)) {
+        if ($protected->has($asset)) {
             return $this->createDeniedResponse();
         }
 
-        // Serve up file response
-        return $this->createResponseFor($filesystem, $asset);
+        // We've looked everywhere and couldn't find a file
+        return $this->createMissingResponse();
+    }
+
+    /**
+     * @return string
+     */
+    private function searchForEquivalentFileID($asset)
+    {
+        // If File is not versionable, let's bail
+        if (!class_exists(Versioned::class) || !File::has_extension(Versioned::class)) {
+            return '';
+        }
+
+        $parsedFileID = $this->parseFileID($asset);
+        if ($parsedFileID && $parsedFileID['Hash']) {
+            // Try to find a live version of this file
+            $stage = Versioned::get_stage();
+            Versioned::set_stage(Versioned::LIVE);
+            $file = File::get()->filter(['FileFilename' => $parsedFileID['Filename']])->first();
+            Versioned::set_stage($stage);
+
+            // If we found a matching live file, let's see if our hash was publish at any point
+            if ($file) {
+                $oldVersionCount = $file->allVersions(
```

Comment:
Could we shortcut this and only run the check if the *current* live version (`$file`) doesn't have this `FileHash`? That'll be the most common scenario, right?

Reply:
That wouldn't make a difference. If you try to access the current live file using the regular hash URL, you will have tripped this condition in `getResponseFor()` and you will never reach this point:
```
        if ($public->has($asset)) {
            return $this->createResponseFor($public, $asset);
        }
```